torch/distributed/_checkpoint/state_dict_loader.py

```
@@ -0,0 +1,234 @@
+            shard_md_from_storage = storage_md.shard_metadata
+            assert shard_md_from_storage is not None
+            # this is a naive quadratic algo that can later be optimized by
+            #   sorting metadata and the shards md
+            # FIXME what does it mean for offset > 0?
+            assert (
+                storage_md.offset == 0
+            ), f"Cannot handle shard '{shard_md}': offset is non-zero"
+
+            # do they overlap?
+            if not _check_shard_metadata_pair_overlap(shard_md, shard_md_from_storage):
+                continue
+
+            shard_volume = 1
+            for (_, _, _, length,) in _shards_get_overlap_region_wrt_saved_tensor(
+                saved_shard=shard_md_from_storage, current_shard=shard_md
+            ):
+                shard_volume *= length
+            read_volume += shard_volume
+
+        shard_volume = 1
+        for size in shard_md.shard_sizes:
+            shard_volume *= size
+        if read_volume != shard_volume:
+            res.append(
+                f"Shard {shard_md} only has {read_volume} available" +
+                "elements but needs {shard_volume}"
+            )
+    return res
+
+
+def validate_metadata(
+    state_dict: Dict[str, Any], metadata: Metadata
+) -> Optional[List[str]]:
+    """
+    Verify if it's possible to correctly load `state_dict` from `metadata`.
+
+    This method can be used to validate if a checkpoint is usable with a given model.
+
+    Sample Code
+    ```
+        my_model: torch.nn.Model = ....
+        my_reader: torch.distributed._checkpoint.StorageReader = ...
+
+        res = torch.distributed._checkpoint.validate_metadata(my_model.state_dict(), my_reader.read_metadata())
+    ```
+    Args:
+        state_dict: A state_dict to verify if it's loadable.
+        metadata: Checkpoint metadata to verify against.
+
+    Returns:
+        None if no issue was found or a List[str] of issues.
```

We should probably just throw an appropriate RuntimeError instead of returning a string of errors like this.